File: UserClientJava/app/src/main/java/com/example/userclientjava/ApiHandler.java
package com.example.userclientjava;

import static java.lang.Thread.sleep;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Arrays;
import java.util.List;

public class ApiHandler {
    List<Coordinates> vehicles;
    List<Coordinates> lampposts;

    public ApiHandler(List<Coordinates> vehicles, List<Coordinates> lampposts) throws MalformedURLException {
        this.vehicles = vehicles;
        this.lampposts = lampposts;
        new Thread(new getVehicles()).start();
        new Thread(new getLampposts()).start();
    }

    List<ApiOnFire> onfireEvent;

    public void addListener(ApiOnFire listener) {
        onfireEvent.add(listener);
    }

    public void setIp(String ip) {
        this.ip = ip;
    }

    String ip = "127.0.0.1";
    String baseUrl = "http://" + ip + ":9696";

    static boolean isActive = false;

    class getVehicles implements Runnable {
        URL url = new URL(baseUrl + "/vehicles");

        getVehicles() throws MalformedURLException {
        }

        @Override
        public void run() {
            while (isActive)
                try {
                    HttpURLConnection con = (HttpURLConnection) url.openConnection();
                    InputStream in = con.getInputStream();
                    InputStreamReader reader = new InputStreamReader(in);
                    char[] buffer = new char[1024];
                    int len = reader.read(buffer);
                    parseVehicles(Arrays.toString(buffer));
                    con.disconnect();
                    for (ApiOnFire i : onfireEvent) i.OnFire();
                    sleep(1000);
                } catch (IOException | InterruptedException e) {
                    throw new RuntimeException(e);
                }
        }
    }

    class getLampposts implements Runnable {
        URL url = new URL(baseUrl + "/lampposts");

        getLampposts() throws MalformedURLException {
        }

        @Override
        public void run() {
            while (isActive)
                try {
                    HttpURLConnection con = (HttpURLConnection) url.openConnection();
                    InputStream in = con.getInputStream();
                    InputStreamReader reader = new InputStreamReader(in);
                    char[] buffer = new char[1024];
                    int len = reader.read(buffer);
                    parseLampposts(Arrays.toString(buffer));
                    con.disconnect();
                    for (ApiOnFire i : onfireEvent) i.OnFire();
                    sleep(1000);
                } catch (IOException | InterruptedException e) {
                    throw new RuntimeException(e);
                }
        }
    }

    private void parseVehicles(String response) {
        String[] tmp = response.replace("[", "").replace("]", "").split("},");
        for (int i = 0; i < tmp.length; i++) {
            tmp[i] = tmp[i].replace("{", "").replace("}", "").replace("\n", "");
        }
        vehicles.clear();
        for (String s : tmp) {
            String[] parts = s.split(",");
            String type = parts[0].split(":")[1].replace("\"", "").replace(" ", "");
            int x = Math.round(Float.parseFloat(parts[1].split(":")[1].replace("\"", "")));
            int y = Math.round(Float.parseFloat(parts[2].split(":")[1].replace("\"", "")));
            String id = parts[3].split(":")[1].replace("\"", "");
            vehicles.add(new Coordinates(CordType.valueOf(type), x, y, id));
        }
    }

    private void parseLampposts(String response) {
        String[] tmp = response.replace("[", "").replace("]", "").split("},");
        for (int i = 0; i < tmp.length; i++) {
            tmp[i] = tmp[i].replace("{", "").replace("}", "").replace("\n", "");
        }
        lampposts.clear();
        for (String s : tmp) {
            String[] parts = s.split(",");
            String type = parts[0].split(":")[1].replace("\"", "").replace(" ", "");
            int x = Math.round(Float.parseFloat(parts[1].split(":")[1].replace("\"", "")));
            int y = Math.round(Float.parseFloat(parts[2].split(":")[1].replace("\"", "")));
            String id = parts[3].split(":")[1].replace("\"", "");
            lampposts.add(new Coordinates(CordType.valueOf(type), x, y, id));
        }
    }
}

File: UserClient/app/src/main/java/com/pacarius/client/ApiHandler.kt
package com.pacarius.client

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import kotlin.math.roundToInt

class ApiHandler(val lampposts: MutableList<Coordinates>, val vehicles: MutableList<Coordinates>, ) {
    init {
//        getLocations()
        getLampposts()
        getVehicles()
    }
    public var OnApiFire : () -> Unit = {}
    public var ip = "192.168.196.240"
    fun getVehicles() {
        CoroutineScope(Dispatchers.IO).launch {
            while (isActive) {
                val url = URL("http://$ip:9696/vehicles")
                try {
                    val con = url.openConnection() as HttpURLConnection
                    val inputStream = con.inputStream
                    val reader = BufferedReader(InputStreamReader(inputStream))
                    val response = reader.readText()
                    parseVehicles(response)
                    con.disconnect()
                } catch (e: Exception) {
                    delay(1000)
                }
                OnApiFire()
                delay(1000)
            }
        }
    }
    private fun getLocations() {
        CoroutineScope(Dispatchers.IO).launch {
            while (isActive) {
                try {
                    val url = URL("http://$ip:9696/vehicles")
                    val con = url.openConnection() as HttpURLConnection
                    try {
                        val inputStream = con.inputStream
                        val reader = BufferedReader(InputStreamReader(inputStream))
                        val response = reader.readText()
                        parseVehicles(response)
                    } finally {
                        con.disconnect()
                    }
                    val url2 = URL("http://$ip:9696/lampposts")
                    val con2 = url2.openConnection() as HttpURLConnection
                    try {
                        val inputStream = con2.inputStream
                        val reader = BufferedReader(InputStreamReader(inputStream))
                        val response = reader.readText()
                        parseLampposts(response)
                    } finally {
                        con2.disconnect()
                    }
                    OnApiFire()
                    delay(1000)
                } catch (e: Exception) {
                    delay(1000)
                }

            }
        }
    }
    fun getLampposts() {
        CoroutineScope(Dispatchers.IO).launch {
            val url = URL("http://$ip:9696/lampposts")
            try {
                val con = url.openConnection() as HttpURLConnection
                val inputStream = con.inputStream
                val reader = BufferedReader(InputStreamReader(inputStream))
                val response = reader.readText()
                parseLampposts(response)
                con.disconnect()
            } catch (e: Exception) {
                delay(1000)
            }

        }
    }

    private fun parseLampposts(response: String) {
        val tmp = response.removeSurrounding("[", "]").split("},")
            .map {
                it.replace("{", "").replace("}", "")
                    .replace("\n", "")
            }
        val a = mutableListOf<Coordinates>()
        val tmp2 = tmp.map { it.split(",") }
        for (i in tmp2) {
            val type = i[0].split(":")[1].replace("\"", "").replace(" ", "")
            val x = i[1].split(":")[1].replace("\"", "").toFloat().roundToInt()
            val y = i[2].split(":")[1].replace("\"", "").toFloat().roundToInt()
            val id = i[3].split(":")[1].replace("\"", "")
            a.add(Coordinates(CoordType.valueOf(type), x, y, id))
        }
        lampposts.addAll(a)
    }
    private fun parseVehicles(response: String){
        val tmp = response.removeSurrounding("[", "]").split("},")
            .map {
                it.replace("{", "").replace("}", "")
                    .replace("\n", "")
            }
        vehicles.clear()
        val tmp2 = tmp.map { it.split(",") }
        for (i in tmp2) {
            val type = i[0].split(":")[1].replace("\"", "").replace(" ", "")
            val x = i[1].split(":")[1].replace("\"", "").toFloat().roundToInt()
            val y = i[2].split(":")[1].replace("\"", "").toFloat().roundToInt()
            val id = i[3].split(":")[1].replace("\"", "")
            vehicles.add(Coordinates(CoordType.valueOf(type), x, y, id))
        }
    }
}